Add support to lock/unlock lba range

paulyoong opened the pull request:
The SPDK repo that this change points at still needs to be determined.

tiagolobocastro committed:
rebuild: Move rebuild task list from the nexus into a global scope (#149)

This makes it easier to interact with the rebuild task list using its destination alone as a key
Also tidy up some of the doc comments and reorder the code a bit

also bring back CI yaml file since we've rebased master off of openebs/master

Commit:
Add pause/resume rebuild support (#158)

tiagolobocastro committed:
Rename RebuildTask to RebuildJob and move rebuild code...

... to a separate file to facilitate mainline rebases
(plus we can always change it back when we integrate into master)

tiagolobocastro committed:
rebuild: run n async copy tasks at a time (#162)

a management async tasks schedules all copy tasks, awaits them and 
schedules enough tasks until the bdev is fully rebuilt
(still using a single reactor this time)

Commit:
Rebuild: Start rebuild when setting child online (#164)

Start a rebuild when setting child online.

- When a child is set offline, stop any ongoing rebuilds for that child
- When setting a child online, set its state to faulted and initiate a rebuild
- Add support for offline and online operations to the "child_operation"

tiagolobocastro committed:
rebuild: remove the teamrebuild CI

remove the test yml as it is not required anymore

tiagolobocastro committed:
rebuild: clean up rebuild job on child removal

when a child is being remove we check for any existing jobs that has:
1. child as dest
in this case we simply stop the job and wait for it to finish
2. child as src
same as above but we also need to start a new rebuild job to rebuild the
dest child.

Atm the wait infra is not here so we have to rely on the child destroy
to block until all the bdevs are closed.
This will be addressed once cas-194 is resolved.

CAS-126

tiagolobocastro committed:
rebuild: add missing_docs attribute

various bits of the rebuild have now been documented
any missing doc will now be a compiler warning

CAS-201

tiagolobocastro committed:
move rebuild and replica into replica dir

tiagolobocastro committed:
rebuild: split rebuild module

splits the rebuild module into impl and api
no functional changes
this should help contain the next set of coming changes
todo: completely hide the rebuild job and expose the trait only

tiagolobocastro committed:
rebuild: split the rebuild state logic

Fixes the rebuild state by introducing and current and a pending state
the pending state is set by the nexus whenever it wants to perform
an operation on the rebuild. The pending state is reconciled when the
job's runner adjusts to the pending state, eg running -> paused

The states are now fully validated by a state machine. No catch all
"_ => ()", was used so the compiler should complain if a new state is
added.

If a pending state is "pending" then a new operation is rejected until
the previous completes.
Internal operations are allowed to override the pending state

Started to make the trait a first class citizen

bonus: fully implement the child src removal by waiting on the rebuild
job to complete before creating a new job to replace the removed one.
This means we now enable the commented out test.
Also fix issues pointed out by semi-standard

CAS-194

tiagolobocastro committed:
rebuild: add progress and map child states

Adds a new get_rebuild_progress rpc to indicate the rebuild progress
Also updates the list_nexus child field:
1. adds a rebuild progress to the child
2. maps the child states to a more simplified version for moac

Also fixed a bug where a non faulted child could self rebuild itself
and added tests for this

CAS-181
CAS-193

tiagolobocastro committed:
Improve the state/status confusion

Avoid "converting" the Nexus and Child states into control plane
specific states by adding a new status for the Nexus and Child
The state now is an internal representation and the status is a higher
level view of the status (a bit like the health) which fits into what
the control plane currently requires.

the Nexus status is now inferred from the its children status:
State               Status
Init       ->       Degraded
Closed     ->       Faulted
Open       ->       All children online: Online
                    At least 1 child online: Degraded
                    All children faulted: Faulted

todo: tidy up the child status logic, including offline/online operation

tiagolobocastro committed:
Automatically rebuild when adding a new child

When a  child is added the control plane expects mayastor to rebuild
the child. Added a new parameter to add_child which allows for this
whilst also allowing for a finer grained control over rebuild.

Adjusted all the required tests for this including semistandard fixes

Added a test to verify that the new parameter works as expected

Resolves CAS-149

Commit:
Support locking/unlocking LBA ranges

This uses SPDK functionality.

Commit:
Add unit tests and format code

Commit:
Amend comment

tiagolobocastro commented on mayastor/src/core/descriptor.rs:
```diff
@@ -109,3 +165,41 @@ impl Debug for Descriptor {
     }
 }
+
+extern "C" fn spdk_cb(
```
can we rename this to a more specific name, like spdk_range_callback or something like that?

paulyoong replied:
Done

Commit:
Use temporary build image

This is required to test against the SPDK changes.

tiagolobocastro commented on mayastor/src/replicas/rebuild/rebuild_api.rs:
```diff
@@ -36,6 +36,26 @@ pub enum RebuildError {
     #[snafu(display("Existing pending state {}", state,))]
     StatePending { state: String },
+    #[snafu(display(
+        "Failed to lock LBA range for blk {}, len {} ",
```
did you mean to add the extra space at the end?

paulyoong replied:
Changed the message

tiagolobocastro commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -205,6 +205,20 @@ impl RebuildJob {
             &mut self.tasks.buffers[id as usize]
         };
 
+        let mut nexus_desc = Bdev::open_by_name(&self.nexus, true).unwrap();
```
should these unwraps be handled?

paulyoong replied:
Done

tiagolobocastro commented on mayastor/tests/lock_lba_range.rs:
```diff
@@ -0,0 +1,230 @@
+};
+use std::sync::{Arc, Mutex};
+
+const NEXUS_NAME: &str = "arbiter_nexus";
```
not using the arbiter now :)

paulyoong replied:
Good catch, changed name

tiagolobocastro commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -205,6 +205,20 @@ impl RebuildJob {
+        let len = copy_buffer.len() as u64 / self.block_size;
+
+        let mut ctx =
+            RangeContext::new(blk, len, nexus_desc.get_channel().unwrap());
```
hmm is the channel going to be acquired and dropped once per segment? If so maybe we should cache it as part of the RebuildJob::new

paulyoong replied:
Done

tiagolobocastro commented on mayastor/tests/lock_lba_range.rs:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    match nexus_create(NEXUS_NAME, NEXUS_SIZE, None, &ch).await {
+        Ok(_) => println!("Created nexus"),
```
do we need these prints, maybe just call unwrap?

paulyoong replied:
Done

tiagolobocastro commented on mayastor/tests/lock_lba_range.rs:
```diff
@@ -0,0 +1,230 @@
+// TODO: Add additional test for issuing front-end I/O then taking a lock
+fn lock_then_fe_io() {
+    test_ini();
+    {
```
do we need the new scope?

paulyoong replied:
If the new scope isn't used, the arc references used prevent the tests from cleaning up properly. Sticking the scope around it ensures the references are dropped before cleanup.

tiagolobocastro commented on mayastor/tests/lock_lba_range.rs:
```diff
@@ -0,0 +1,230 @@
+        let ctx_clone = Arc::clone(&ctx);
+        reactor.send_future(async move {
+            unlock_range(&mut ctx_clone.lock().unwrap()).await;
+            trace!("*** RELEASE LOCK ***");
```
capitalization/wording is different to the previous test, should it be same/similar?

paulyoong replied:
Done

tiagolobocastro reviewed (commented):
Looks good, the only thing of note is the IoChannel comment

Commit:
Address review comments

Commit:
Update comments

Commit:
Add nexus descriptor to rebuild job struct

Commit:
Fix up failing CI builds

Sometimes the rebuild progress doesn't go up because the rebuild doesn't
have enough time to start before it is checked.

Commit:
Add sleeps to allow js tests to pass

Commit:
Remove I/O channel from rebuild job

This removes the need to use Arc.
The I/O channel is allocated when creating the context. Once the context
is out of scope, the I/O channel is no longer required.

Commit:
Provide RangeContext as the callback argument

This allows oneshot channels to be used.

Commit:
Improve error handling when performing rebuild

Refactor to ensure the LBA range is unlocked even if the copy fails.

Commit:
Merge branch 'master' into lock-lba-range

Commit:
Fixup code after merge from master

Commit:
Update comments
